prow/cmd/pipeline/controller.go

```diff
@@ -613,24 +617,39 @@ func reconcile(c reconciler, key string) error {
 		return fmt.Errorf("no pipelinerun found or created for %q, wantPipelineRun was %v", key, wantPipelineRun)
 	}
 	wantState, wantMsg := prowJobStatus(runs)
-	return updateProwJobState(c, pj, wantState, wantMsg)
+	return updateProwJobState(c, pj, wantState, wantMsg, ctx, namespace, runs)
 }
 
-func updateProwJobState(c reconciler, pj *prowjobv1.ProwJob, state prowjobv1.ProwJobState, msg string) error {
+func updateProwJobState(c reconciler, pj *prowjobv1.ProwJob, state prowjobv1.ProwJobState, msg, context, namespace string, runs []*pipelinev1alpha1.PipelineRun) error {
 	haveState := pj.Status.State
 	haveMsg := pj.Status.Description
 	if haveState != state || haveMsg != msg {
 		npj := pj.DeepCopy()
-		if npj.Status.StartTime.IsZero() {
-			npj.Status.StartTime = c.now()
-		}
-		if npj.Status.CompletionTime.IsZero() && finalState(state) {
-			now := c.now()
-			npj.Status.CompletionTime = &now
+		if state == prowjobv1.FailureState && strings.Contains(msg, pipelineRunShouldRetryMessage) {
+			// If the pipeline failed due to the known race condition, wipe the status and retry.
+			npj.Status = prowjobv1.ProwJobStatus{
+				StartTime: c.now(),
+				State:     prowjobv1.TriggeredState,
+			}
+			// Delete the existing PipelineRun(s) so that we don't get confused by them in subsequent reconciliation.
+			for _, r := range runs {
+				logrus.Infof("Delete pipelinerun: %s", r.Name)
+				if err := c.deletePipelineRun(context, namespace, r.Name); err != nil {
+					return fmt.Errorf("delete pipelinerun: %v", err)
```

Comment:
This will lead to a reconciliation retry with the prow job in a different state. The next loop execution most likely won't cleanup the left pipeline runs. Is this error critical for completely the reconciliation loop?

Reply:
Good question...we need to not see the “bad” run(s) on next reconciliation, so that it’ll create a new run and not just treat the old failure(s) as meaning the ProwJob has failed. If there’s a better way to do this than deleting the old run(s), I’m completely open to it.

Reply:
Let's try like this, and see how it works.